### PR TITLE
DateTime field support

### PR DESCRIPTION
**Pull request by evanxg852000**
- Implements the DateTime field support 
- updated tantivy version to include DateTime changes 
Closes https://github.com/quickwit-oss/quickwit/issues/1328

### CODE REVIEW

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 1]
```diff
@@ -0,0 +1,630 @@
...
+            indexed: options.indexed,
+            stored: options.stored,
+            fast: options.fast,
+            parsers,
+        }
+    }
+}
+
+// Parser function types
+type StringDateTimeParser = Arc<dyn Fn(&str) -> Result<OffsetDateTime, String> + Sync + Send>;
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:130]: I don't think those function types are buying us anything. We could just iterate over the list of formats and get away with it. Less code, no dynamic dispatch. That would also remove the need for `QuickwitDateOptionsDeser`:

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 2]
```diff
@@ -0,0 +1,630 @@
...
+    }
+
+    pub fn parse_number(&self, value: i64) -> Result<OffsetDateTime, String> {
+        self.parsers.parse_number(value)
+    }
+}
+
+#[derive(Deserialize)]
+#[serde(deny_unknown_fields)]
+struct QuickwitDateOptionsDeser {
```
[the remark this answers is not in the document]
Comment (evanxg852000) [reply quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:98]: Yes

File: docs/configuration/index-config.md [hunk 3]
```diff
@@ -84,11 +84,11 @@ Today, only the s3 storage is available when running several searcher nodes.
 
 ## Doc mapping
 
-The doc mapping defines how a document and the fields it contains are stored and indexed for a given index. A document is a collection of named fields, each having its own data type (text, binary, i64, u64, f64).
+The doc mapping defines how a document and the fields it contains are stored and indexed for a given index. A document is a collection of named fields, each having its own data type (text, binary, date, i64, u64, f64).
```
[the remark this answers is not in the document]
Comment (evanxg852000) [reply docs/configuration/index-config.md:87]: We could definitely have both: date being based on datetime. For better support (query parsing, display). I think we need to start from tantivy.  Would you suggest renaming the `date` type as `datetime` now?

[the remark this answers is not in the document]
Comment (guilload) [reply docs/configuration/index-config.md:87]: Yes, let's rename everything `Datetime` for now. We can add a proper `Date` type later with better tantivy support.

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+            "iso8601" => Ok(DateTimeFormat::ISO8601),
+            "unix_ts_secs" => Ok(DateTimeFormat::Timestamp(DatePrecision::Seconds)),
+            "unix_ts_millis" => Ok(DateTimeFormat::Timestamp(DatePrecision::Milliseconds)),
+            "unix_ts_micros" => Ok(DateTimeFormat::Timestamp(DatePrecision::Microseconds)),
+            _ => Ok(DateTimeFormat::Strftime(value)),
+        }
+    }
+}
+
+impl Serialize for DateTimeFormat {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:194]: We  may want to implement `as_str` `DateTimeFormat`?

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+            DateTimeFormat::Timestamp(precision) => match precision {
+                DatePrecision::Seconds => serializer.serialize_str("unix_ts_secs"),
+                DatePrecision::Milliseconds => serializer.serialize_str("unix_ts_millis"),
+                DatePrecision::Microseconds => serializer.serialize_str("unix_ts_micros"),
+            },
+        }
+    }
+}
+
+fn date_time_format_to_string(format: &DateTimeFormat) -> String {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:211]: and then implement `Display` for `DateTimeFormat` instead.

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+        DateTimeFormat::Timestamp(precision) => match precision {
+            DatePrecision::Seconds => "unix_ts_secs",
+            DatePrecision::Milliseconds => "unix_ts_millis",
+            DatePrecision::Microseconds => "unix_ts_micros",
+        },
+    };
+    format_str.to_string()
+}
+
+fn date_time_format_from_string(format_str: String) -> DateTimeFormat {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:226]: That's [FromStr](https://doc.rust-lang.org/std/str/trait.FromStr.html#)

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+
+fn default_input_formats() -> IndexSet<DateTimeFormat> {
+    let mut input_formats = IndexSet::new();
+    input_formats.insert(DateTimeFormat::RCF3339);
+    input_formats.insert(DateTimeFormat::Timestamp(DatePrecision::default()));
+    input_formats
+}
+
+/// Converts a timestamp to displayable date time in available formats.
+pub fn timestamp_to_datetime_str(
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:272]:
```suggestion
pub(crate) fn timestamp_to_datetime_str(
```

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+            precision: DatePrecision::default(),
+            indexed: true,
+            stored: true,
+            fast: false,
+        }
+    }
+}
+
+impl QuickwitDateOptions {
+    pub fn parse_string(&self, value: String) -> Result<OffsetDateTime, String> {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:76]:
```suggestion
    pub(crate) fn parse_string(&self, value: String) -> Result<OffsetDateTime, String> {
```

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 9]
```diff
@@ -0,0 +1,539 @@
...
+            "Could not parse date `{}` using the specified formats `{}`.",
+            value,
+            self.input_formats
+                .iter()
+                .map(date_time_format_to_string)
+                .join(", ")
+        ))
+    }
+
+    pub fn parse_number(&self, value: i64) -> Result<OffsetDateTime, String> {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:102]:
```suggestion
    pub(crate) fn parse_number(&self, value: i64) -> Result<OffsetDateTime, String> {
```

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 10]
```diff
@@ -0,0 +1,539 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use chrono::TimeZone;
+use chrono_tz::Tz;
+use derivative::Derivative;
+use indexmap::IndexSet;
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:23]: Nice.

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+#[serde(deny_unknown_fields)]
+pub struct QuickwitDateOptions {
+    #[serde(default)]
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub description: Option<String>,
+
+    /// Accepted input formats.
+    #[serde(default = "default_input_formats")]
+    #[serde(serialize_with = "serialize_input_formats")]
+    #[serde(deserialize_with = "deserialize_input_formats")]
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:45]: `IndexSet` seems to serialize/deserialize without those custom serializer/deserializer. There should be no ordering issue for a set serialized as a list in JSON.

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 12]
```diff
@@ -0,0 +1,539 @@
...
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use chrono::TimeZone;
+use chrono_tz::Tz;
+use derivative::Derivative;
+use indexmap::IndexSet;
+use itertools::Itertools;
+use serde::ser::SerializeSeq;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use tantivy::{DatePrecision, DateTime};
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:27]:
```suggestion
use tantivy::{DatePrecision as DateTimePrecision, DateTime};
```

File: quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs [hunk 13]
```diff
@@ -98,6 +102,19 @@ impl LeafType {
                     Err(format!("Expected bool value, got '{}'.", json_val))
                 }
             }
+            LeafType::Date(options) => {
+                let date_time = match json_val {
+                    JsonValue::String(text) => options.parse_string(text)?,
+                    JsonValue::Number(number) => options.parse_number(number.as_i64().unwrap())?,
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs:108]: Remove the unwrap and return proper error: "Failed to blabla".

File: quickwit-doc-mapper/src/query_builder.rs [hunk 14]
```diff
@@ -30,10 +30,13 @@ pub(crate) fn build_query(
     schema: Schema,
     request: &SearchRequest,
     default_field_names: &[String],
+    // date_time_parsers: HashSet<String, DateTimeParsersHolder>,
```
Comment (guilload) [inline quickwit-doc-mapper/src/query_builder.rs:33]: remove?

File: quickwit-indexing/Cargo.toml [hunk 15]
```diff
@@ -49,11 +49,11 @@ rusoto_kinesis = { version = "0.48", default-features = false, features = [
...
-tantivy = { git = "https://github.com/quickwit-oss/tantivy/", rev = "2406d92", default-features = false, features = [
-    "mmap",
-    "lz4-compression",
-    "zstd-compression",
-    "quickwit",
+tantivy = { git = "https://github.com/quickwit-oss/tantivy/", rev = "d24f31f", default-features = false, features = [
+  "mmap",
+  "lz4-compression",
+  "zstd-compression",
+  "quickwit"
```
Comment (guilload) [inline quickwit-indexing/Cargo.toml:56]: lol, just review a PR from Paul where he added the trailing comma. This makes me feel better about myself, I felt I was the only one OCD on the team.

File: quickwit-directories/src/bundle_directory.rs [hunk 16]
```diff
@@ -86,15 +86,15 @@ fn split_footer(file_slice: FileSlice) -> io::Result<(FileSlice, FileSlice)> {
 
 /// Return two slices for given split: `[body and bundle meta data] [hotcache]`
 pub fn get_hotcache_from_split(data: OwnedBytes) -> io::Result<OwnedBytes> {
-    let split_file = FileSlice::new(Box::new(data));
+    let split_file = FileSlice::new(Arc::new(data));
```
Comment (guilload) [inline quickwit-directories/src/bundle_directory.rs:89]: nit: Next time, update tantivy in a separate PR to reduce the noise.

Comment (evanxg852000) [reply quickwit-directories/src/bundle_directory.rs:89]: ok

File: docs/configuration/index-config.md [hunk 17]
```diff
@@ -172,6 +172,42 @@ fast: true
...
 
+#### `datetime` type
+
+The `datetime` type can accepts multiple formats and a storage precision. The following formats are supported but need to be explicitly requested via configuration.
+- `rfc3339`, `rfc2822`, `iso8601`: Parsing dates using standard specified formats.
+- `strftime`: Parsing dates using the Unix [strftime](https://man7.org/linux/man-pages/man3/strftime.3.html) format.
+- `unix_ts_secs`, `unix_ts_millis`, `unix_ts_micros`: Parsing dates from numbers (timestamp). Only one can be used in configuration. `unix_ts_secs` is added to the list by default if none is specified. 
+
+:::info
+When accepting multiple formats, the corresponding parsers are tried in order they are declared.
```
Comment (guilload) [inline docs/configuration/index-config.md:183]:
```suggestion
When specifying multiple input formats, the corresponding parsers are tried in the order they are declared. 
```

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 18]
```diff
@@ -0,0 +1,512 @@
...
+            self.input_formats
+                .iter()
+                .map(ToString::to_string)
+                .join(", ")
+        ))
+    }
+
+    pub(crate) fn parse_number(&self, value: i64) -> Result<OffsetDateTime, String> {
+        for format in self.input_formats.iter() {
+            match format {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:104]:
```rust
if let DateTimeFormat::Timestamp(precision) = format {
    return unix_timestamp_parse(value, precision)
}
```

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 19]
```diff
@@ -0,0 +1,512 @@
...
+    } else {
+        chrono::NaiveDateTime::parse_from_str(value, format).map_err(|error| error.to_string())?
+    };
+
+    OffsetDateTime::from_unix_timestamp_nanos(date_time.timestamp_nanos() as i128)
+        .map_err(|error| error.to_string())
+}
+
+/// Recognizes numbers as unix timestamp with a precision.
+fn unix_timestamp_parse(
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:148]:
```suggestion
fn parse_unix_timestamp(
```

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 20]
```diff
@@ -0,0 +1,512 @@
...
+}
+
+impl DateTimeFormat {
+    pub fn as_str(&self) -> &str {
+        match self {
+            DateTimeFormat::RCF3339 => "rfc3339",
+            DateTimeFormat::RFC2822 => "rfc2822",
+            DateTimeFormat::ISO8601 => "iso8601",
+            DateTimeFormat::Strftime(format) => format,
+            DateTimeFormat::Timestamp(precision) => match precision {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:186]:
```rust
DateTimeFormat::Timestamp(DateTimePrecision::Seconds) => "unix_ts_secs"
...
```

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 21]
```diff
@@ -0,0 +1,512 @@
...
+                DateTimePrecision::Seconds => "unix_ts_secs",
+                DateTimePrecision::Milliseconds => "unix_ts_millis",
+                DateTimePrecision::Microseconds => "unix_ts_micros",
+            },
+        }
+    }
+}
+
+impl Display for DateTimeFormat {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:196]: The codebase aligns on
```rust
fn fmt(&self, f: &mut std::fmt::Formatter)
```

File: quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs [hunk 22]
```diff
@@ -0,0 +1,512 @@
...
+}
+
+/// Parses DateTime strings using default ISO8601 formatting.
+/// Examples: 2010-11-21T09:55:06.000000000+02:00, 2010-11-12 9:55:06 +2:00
+fn iso8601_parse(value: &str) -> Result<OffsetDateTime, String> {
+    OffsetDateTime::parse(value, &Iso8601::DEFAULT).map_err(|error| error.to_string())
+}
+
+/// Parses DateTime strings using the unix strftime formatting.
+fn strftime_parse(value: &str, format: &str) -> Result<OffsetDateTime, String> {
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:134]:
```suggestion
fn parse_strftime(value: &str, format: &str) -> Result<OffsetDateTime, String> {
```

Comment (guilload) [reply quickwit-doc-mapper/src/default_doc_mapper/date_time_type.rs:134]: The codebase aligns on putting the verb first.

File: quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs [hunk 23]
```diff
@@ -172,6 +194,14 @@ impl MappingLeaf {
         let json_type = self.typ.json_type();
         if let Some(json_val) = extract_json_val(json_type, named_doc, field_path, self.cardinality)
         {
+            if let (LeafType::DateTime(options), Some(timestamp)) =
+                (self.get_type(), json_val.as_i64())
+            {
+                let date_time_str =
+                    timestamp_to_datetime_str(timestamp, &options.precision).unwrap();
```
Comment (guilload) [inline quickwit-doc-mapper/src/default_doc_mapper/mapping_tree.rs:201]: This is probably an expect, right?

**Review by guilload — APPROVED** [review]

**Pull request merged**